RecoMuon/TrackerSeedGenerator/plugins/TSGForOIDNN.cc

```diff
@@ -0,0 +1,974 @@
+  if (getStrategyFromDNN_) {
+    edm::FileInPath dnnMetadataPath(dnnMetadataPath_);
+    pt::read_json(dnnMetadataPath.fullPath(), metadata);
+    tensorflow::setLogging("2");
+
+    if (useRegressor_) {
+      // use regressor
+      dnnModelPath_HB_ = metadata.get<std::string>("HB.dnnmodel_path");
+      edm::FileInPath dnnPath_HB(dnnModelPath_HB_);
+      graphDef_HB_ = tensorflow::loadGraphDef(dnnPath_HB.fullPath());
+      tf_session_HB_ = tensorflow::createSession(graphDef_HB_);
+
+      dnnModelPath_HLIP_ = metadata.get<std::string>("HLIP.dnnmodel_path");
+      edm::FileInPath dnnPath_HLIP(dnnModelPath_HLIP_);
+      graphDef_HLIP_ = tensorflow::loadGraphDef(dnnPath_HLIP.fullPath());
+      tf_session_HLIP_ = tensorflow::createSession(graphDef_HLIP_);
+
+      dnnModelPath_HLMuS_ = metadata.get<std::string>("HLMuS.dnnmodel_path");
+      edm::FileInPath dnnPath_HLMuS(dnnModelPath_HLMuS_);
+      graphDef_HLMuS_ = tensorflow::loadGraphDef(dnnPath_HLMuS.fullPath());
+      tf_session_HLMuS_ = tensorflow::createSession(graphDef_HLMuS_);
+    } else {
+      // use classifier (default)
+      dnnModelPath_barrel_ = metadata.get<std::string>("barrel.dnnmodel_path");
+      edm::FileInPath dnnPath_barrel(dnnModelPath_barrel_);
+      graphDef_barrel_ = tensorflow::loadGraphDef(dnnPath_barrel.fullPath());
+      tf_session_barrel_ = tensorflow::createSession(graphDef_barrel_);
+
+      dnnModelPath_endcap_ = metadata.get<std::string>("endcap.dnnmodel_path");
+      edm::FileInPath dnnPath_endcap(dnnModelPath_endcap_);
+      graphDef_endcap_ = tensorflow::loadGraphDef(dnnPath_endcap.fullPath());
+      tf_session_endcap_ = tensorflow::createSession(graphDef_endcap_);
+    }
+  }
+  produces<std::vector<TrajectorySeed> >();
+}
+
+TSGForOIDNN::~TSGForOIDNN() {
+  if (getStrategyFromDNN_) {
+    if (useRegressor_) {
+      tensorflow::closeSession(tf_session_HB_);
+      tensorflow::closeSession(tf_session_HLIP_);
+      tensorflow::closeSession(tf_session_HLMuS_);
+      delete graphDef_HB_;
+      delete graphDef_HLIP_;
+      delete graphDef_HLMuS_;
+    } else {
+      tensorflow::closeSession(tf_session_barrel_);
+      tensorflow::closeSession(tf_session_endcap_);
+      delete graphDef_barrel_;
+      delete graphDef_endcap_;
+    }
+  }
+}
+
+//
+// Produce seeds
+//
+void TSGForOIDNN::produce(edm::StreamID sid, edm::Event& iEvent, edm::EventSetup const& iEventSetup) const {
+  // Initialize variables
+  unsigned int numSeedsMade = 0;
+  unsigned int layerCount = 0;
+  unsigned int hitlessSeedsMadeIP = 0;
+  unsigned int hitlessSeedsMadeMuS = 0;
+  unsigned int hitSeedsMade = 0;
+  unsigned int hitDoubletSeedsMade = 0;
+
+  // Surface used to make a TSOS at the PCA to the beamline
+  Plane::PlanePointer dummyPlane = Plane::build(Plane::PositionType(), Plane::RotationType());
+
+  // Read ESHandles
+  edm::ESHandle<Chi2MeasurementEstimatorBase> estimatorH = iEventSetup.getHandle(t_estimatorH_);
```

Comment:
```suggestion
  const Chi2MeasurementEstimatorBase& estimator = iEventSetup.getData(t_estimatorH_);
```
same for the rest; unless there is a clear case for using the handle

Reply:
There is `magfieldH.product()`, so I'm not sure how to get around it without using the handle.
I was able to remove all other handles without issues.

Reply:
Resolved